lib/users.js

```diff
@@ -484,40 +485,85 @@ Users.prototype._getAllGroupsADCached = function (profile, callback) {
+  return new Promise(function(resolve, reject) {
+    client.search(baseGroups, opts, function(err, ldapEvents) {
+      if (err) {
+        return reject(err);
+      }
+
+      let entries = [];
+      ldapEvents.on('searchEntry', (entry) => {
+        entries.push(entry.object);
+      });
+
+      ldapEvents.on('error', (err) => {
+        return resolve(entries);
+      });
+
+      ldapEvents.on('end', (result) => {
+        if (result.status !== 0) {
+          console.log('Unexpected result on _getAllGroupsAD', result.status);
+        }
+        return resolve(entries);
+      });
+    });
+  });
+}
+
+Users.prototype._getAllGroupsAD = function(dn, callback) {
   var self = this;
-  var opts = {
-    scope: 'sub',
-    filter: nconf.get('LDAP_SEARCH_GROUPS').replace(/\{0\}/ig, dn),
-    attributes: _.uniq(['cn', 'dn', 'memberOf'].concat(nconf.get('GROUP_PROPERTIES'))),
-    timeLimit: nconf.get('GROUPS_TIMEOUT_SECONDS'),
-    derefAliases: nconf.get('GROUPS_DEREF_ALIASES')
-  };
 
   callback = cb(callback)
     .timeout(nconf.get('GROUPS_TIMEOUT_SECONDS') * 1000)
     .once();
 
-  self._client.search(self._baseGroups, opts, function (err, res) {
-    if (err) {
-      return callback(err);
+  async.parallel([
+    async function(cb) {
+      try {
+        console.info('gathering primary groups for', dn);
+        let primaryGroups = await findGroups(self._client, self._baseGroups, {
+          scope: 'sub',
+          filter: nconf.get('LDAP_SEARCH_PRIMARY_GROUPS').replace(/\{0\}/ig, dn),
+          attributes: _.uniq(['cn', 'dn', 'memberOf'].concat(nconf.get('GROUP_PROPERTIES'))),
+          timeLimit: nconf.get('GROUPS_TIMEOUT_SECONDS'),
+          derefAliases: nconf.get('GROUPS_DEREF_ALIASES')
+        });
+        return cb(null, primaryGroups);
+      } catch (err) {
+        console.error('failed during primary group search', err.message);
+        return cb(err);
+      }
+    },
+    async function(cb) {
+      try {
+        console.info('gathering secondary groups for', dn);
+        let secondaryGroups = await findGroups(self._client, self._baseGroups, {
+          scope: 'sub',
+          filter: nconf.get('LDAP_SEARCH_GROUPS').replace(/\{0\}/ig, dn),
+          attributes: _.uniq(['cn', 'dn', 'memberOf'].concat(nconf.get('GROUP_PROPERTIES'))),
+          timeLimit: nconf.get('GROUPS_TIMEOUT_SECONDS'),
+          derefAliases: nconf.get('GROUPS_DEREF_ALIASES')
+        });
+        return cb(null, secondaryGroups);
+      } catch (err) {
+        console.error('failed during secondary group search', err.message);
+        return cb(err);
+      }
     }
-    var entries = [];
-
-    res.on('searchEntry', function (entry) {
-      entries.push(entry.object);
-    });
-
-    res.on('error', function (err) {
+  ],
+  function(err, results) {
+    let entries = _.flatten(results);
+    if (err) {
       console.log('search groups', err.message);
-      callback(null, entries);
-    });
+      return callback(null, entries);
```

Comment:
This matches the current behavior which is to always return the entries list, which is typically an empty array if there is a failure. I suspect this may be to deal with non-AD connections, but I haven't tested this theory yet.